### PR TITLE
Split RGB Matrix

### PR DESCRIPTION
**Pull request by XScorpion2**
## Description

Final implementation of the Split RGB Matrix feature. This builds upon the previous features of #11046  and #10997 

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: tmk_core/common/keyboard.c
Comment (drashna) [inline tmk_core/common/keyboard.c:357]: I'm not a fan of moving this.  

The point of `should_process_keypress` is that it can be over-ridden.  And if we need to do so for any reason, it should be overwritten.  Eg, something like this, in split_util.c would make more sense, IMO:

```c
bool should_process_keypress(void) { 
    is_keyboard_master(); 
    return true;
}
```

And then below, replace `should_process_keypress()` with `is_keyboard_master()`. 

```suggestion
    if (should_process_keypress()) {
        for (uint8_t r = 0; r < MATRIX_ROWS; r++) {
            matrix_row    = matrix_get_row(r);
            matrix_change = matrix_row ^ matrix_prev[r];
            if (matrix_change) {
#ifdef MATRIX_HAS_GHOST
                if (has_ghost_in_row(r, matrix_row)) {
                    continue;
                }
#endif
                if (debug_matrix) matrix_print();
                matrix_row_t col_mask = 1;
                for (uint8_t c = 0; c < MATRIX_COLS; c++, col_mask <<= 1) {
                    if (matrix_change & col_mask) {
                         if (is_keyboard_master()) {

```

File: tmk_core/common/keyboard.c
Comment (drashna) [inline tmk_core/common/keyboard.c:366]: I feel the naming for this function should be changed, since it's not longer part of the `process_*` chain. 

But, `rgb_matrix_matrix_handler` sounds like a horrendous name, and I can't think of a better one.

File: tmk_core/common/keyboard.c
Comment (drashna) [inline tmk_core/common/keyboard.c:397]: (probably should move the init call to keyboard.c too, for consistency)

Comment (XScorpion2) [reply tmk_core/common/keyboard.c:397]: solid point

File: tmk_core/common/keyboard.c
Comment (XScorpion2) [reply tmk_core/common/keyboard.c:366]: rgb_matrix_handle_matrix?

File: tmk_core/common/keyboard.c
Comment (XScorpion2) [reply tmk_core/common/keyboard.c:357]: The point of moving this is that this changed setup allows for systems that need to react to electrical matrix changes can do so without also triggering the entire key process loop. So should_process_keypress remains as overridable to allow key events to trigger, and that is independent from what rgb matrix needs to do.

File: tmk_core/common/eeconfig.h
Comment (drashna) [inline tmk_core/common/eeconfig.h:53]: Since I can't tag it, please increment `EECONFIG_MAGIC_NUMBER` as well.   This way, the shifting EEPROM ranges don't cause weird issues.  (eg, this brute force resets the eeprom to default values)

File: tmk_core/common/keyboard.c
Comment (drashna) [reply tmk_core/common/keyboard.c:366]: Sounds just as horrible as my suggestion.  somebody may have a better name, but... for now, I think it's okay.

File: tmk_core/common/keyboard.c
Comment (drashna) [reply tmk_core/common/keyboard.c:357]: Well, in that case, wouldn't it make more sense to remove the ifdef for "oled task". and always have `matrix_scan`s status be dumped to a boolean?  And then check that, in the case for stuff like this. 

eg: 
```c
                            if (is_keyboard_master() 
#if defined(RGB_MATRIX_SPLIT)
                                || changed
#endif
                                  ) { 
```

Comment (XScorpion2) [reply tmk_core/common/keyboard.c:357]: Not really, as in RGB Matrix's case it needs to react per key, not per matrix.

Comment (XScorpion2) [reply tmk_core/common/keyboard.c:357]: The way I see how features need to react to changes in state basically falls into 4 categories: 
- Any change in state (OLED inactivity timer, RGB inactivity timer, etc)
- Any changes per key (Key reactive RGB effects)
- Any keycodes (process_***_task)
- Custom

File: tmk_core/common/eeconfig.h
Comment (XScorpion2) [reply tmk_core/common/eeconfig.h:53]: fixed

File: tmk_core/common/keyboard.c
Comment (XScorpion2) [reply tmk_core/common/keyboard.c:397]: fixed

**Comment by XScorpion2** [comment]
rebased & force pushed

**Comment by BenRoe** [comment]
@XScorpion2 i added your split_rgb_matrix to my wip split keyboard pcb and it works.
https://github.com/BenRoe/qmk_firmware/tree/split_rgb_matrix_2020/keyboards/keycapsss/kimiko

One thing is not working. Some animations move from right-half (slave) to left-half(master), but not the other way arround.
[Video](https://i.imgur.com/BdLvAXt.mp4)

**Comment by XScorpion2** [comment]
> 
> 
> @XScorpion2 i added your split_rgb_matrix to my wip split keyboard pcb and it works.
> https://github.com/BenRoe/qmk_firmware/tree/split_rgb_matrix_2020/keyboards/keycapsss/kimiko
> 
> One thing is not working. Some animations move from right-half (slave) to left-half(master), but not the other way arround.
> [Video](https://i.imgur.com/BdLvAXt.mp4)

Sounds like you didn't flash both halves with the updated firmware.

**Comment by BenRoe** [comment]
> Sounds like you didn't flash both halves with the updated firmware.
Flashed your branch firmware again and still the same problem.

**Comment by XScorpion2** [comment]
> 
> 
> > Sounds like you didn't flash both halves with the updated firmware.
> 
> Flashed your branch firmware again and still the same problem.

Did you enable SPLIT_TRANSPORT_MIRROR?

**Comment by BenRoe** [comment]
> Did you enable SPLIT_TRANSPORT_MIRROR?

That was the missing part. Now every animation works.
Have to test it more, but it seems that this PR is working without problems.
Looking forward to be merged.

**Comment by nicsuzor** [comment]
Thanks for this! I'm running split_common on ARM on crkbd, and I tried to test this, but I couldn't get the two halves to talk. I keep getting the errors `serial::NO_RESPONSE` and `serial::FAIL` from soft_serial_transaction() in serial.c. I confirm I flashed both sides and set `SPLIT_TRANSPORT_MIRROR` and `RGB_MATRIX_SPLIT`. 

I'm sorry, I'm likely missing something obvious, but do I have to make any other changes to my config? Without these changes, SPLIT_TRANSPORT_MIRROR works fine.

**Comment by BenRoe** [comment]
@nicsuzor the CRKBD has it's own custom split rgb.
You can test it with the keymap from @soundmonster 

```
make crkbd:soundmonster
```

**Comment by zvecr** [comment]
Actually the issue is more with using bitbang transport, which is not advised with rgb. Though from memory it did work with the pwm rgb driver.

**Comment by drashna** [comment]
In testing, I have found that this breaks rgb matrix suspend/sleep. It fails to clear the LEDs.  Not sure why, though.

**Comment by drashna** [comment]
Testing this, it seems to work well, even without mirroring the matrix.

**Comment by XScorpion2** [comment]
> 
> 
> Testing this, it seems to work well, even without mirroring the matrix.

I'll have the common function implemented later this weekend to wrap this all up nicely. Plus rebase these changes.

**Comment by tzarc** [comment]
Some funky stuff going on:

Adding RGB_MATRIX to the Djinn ends up having semi-working reactive effects
 * master->slave sync seems ok
 * slave->master sync doesn't seem to do anything

**However, the bigger problem is that any keypress on the slave isn't registered at all.**

Furthermore, building Djinn with its old RGBLIGHT config (i.e. no RGB_MATRIX enabled) gives linking errors:
```
 | /usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/bin/ld: .build/obj_handwired_tzarc_djinn_tzarc/quantum/split_common/transport.a(transport.o): in function `transport_rgblight_master':
 | /home/nickb/qmk_build/qmk_firmware/quantum/split_common/transport.c:374: undefined reference to `rgb_matrix_config'
 | /usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/bin/ld: .build/obj_handwired_tzarc_djinn_tzarc/quantum/split_common/transport.a(transport.o): in function `transport_rgblight_slave':
 | /home/nickb/qmk_build/qmk_firmware/quantum/split_common/transport.c:317: undefined reference to `rgb_matrix_config'
 | collect2: error: ld returned 1 exit status
```

**Comment by drashna** [comment]
Oh, one weird artifact, `rgb_matrix_task` needs to be manually called by the suspend code, or it doesn't "clean up" the animations/matrix.  It did so before, because the suspend code calls matrix_scan, but now that doesn't work.

File: quantum/rgb_matrix.c
Comment (tzarc) [inline quantum/rgb_matrix.c:147]:
```suggestion
#if defined(RGB_MATRIX_ENABLE) && defined(RGB_MATRIX_SPLIT)
```

File: quantum/rgb_matrix.c
Comment (tzarc) [inline quantum/rgb_matrix.c:191]:
```suggestion
#if defined(RGB_MATRIX_ENABLE) && defined(RGB_MATRIX_SPLIT)
```

File: quantum/rgb_matrix.c
Comment (tzarc) [inline quantum/rgb_matrix.c:200]:
```suggestion
#if defined(RGB_MATRIX_ENABLE) && defined(RGB_MATRIX_SPLIT)
```

File: quantum/rgb_matrix.c
Comment (tzarc) [inline quantum/rgb_matrix.c:209]:
```suggestion
#if !(defined(RGB_MATRIX_ENABLE) && defined(RGB_MATRIX_SPLIT))
```

File: quantum/split_common/transport.c
Comment (tzarc) [inline quantum/split_common/transport.c:25]:
```suggestion
#if defined(RGB_MATRIX_ENABLE) && defined(RGB_MATRIX_SPLIT)
```

File: quantum/split_common/transport.c
Comment (tzarc) [inline quantum/split_common/transport.c:61]:
```suggestion
#    if defined(RGB_MATRIX_ENABLE) && defined(RGB_MATRIX_SPLIT)
```

File: quantum/split_common/transport.c
Comment (tzarc) [inline quantum/split_common/transport.c:152]:
```suggestion
#    if defined(RGB_MATRIX_ENABLE) && defined(RGB_MATRIX_SPLIT)
```

File: quantum/split_common/transport.c
Comment (tzarc) [inline quantum/split_common/transport.c:202]:
```suggestion
#    if defined(RGB_MATRIX_ENABLE) && defined(RGB_MATRIX_SPLIT)
```

File: quantum/split_common/transport.c
Comment (tzarc) [inline quantum/split_common/transport.c:245]:
```suggestion
#    if defined(RGB_MATRIX_ENABLE) && defined(RGB_MATRIX_SPLIT)
```

File: quantum/split_common/transport.c
Comment (tzarc) [inline quantum/split_common/transport.c:366]:
```suggestion
#    if defined(RGB_MATRIX_ENABLE) && defined(RGB_MATRIX_SPLIT)
```

File: quantum/split_common/transport.c
Comment (tzarc) [inline quantum/split_common/transport.c:409]:
```suggestion
#    if defined(RGB_MATRIX_ENABLE) && defined(RGB_MATRIX_SPLIT)
```

**Review by tzarc — CHANGES_REQUESTED** [review]
These changes should hopefully fixup building RGBLIGHT when there's an `RGB_MATRIX_SPLIT` present in `config.h` -- i.e. allowing for both to be used without modification.

File: quantum/split_common/matrix.c
Comment (drashna) [inline quantum/split_common/matrix.c:260]:
```suggestion
        if (!transport_master(matrix + thisHand, slave_matrix)) {
```

Comment (tzarc) [reply quantum/split_common/matrix.c:260]: @drashna's suggestion fixes my previous issue with the slave not reporting keypresses.

**Comment by XScorpion2** [comment]
Will resolve all these issues this weekend

File: quantum/split_common/matrix.c
Comment (XScorpion2) [reply quantum/split_common/matrix.c]: fixed in base pr: https://github.com/qmk/qmk_firmware/pull/11046

File: quantum/rgb_matrix.c
Comment (XScorpion2) [reply quantum/rgb_matrix.c]: Fixed

File: quantum/rgb_matrix.c
Comment (XScorpion2) [reply quantum/rgb_matrix.c]: Fixed

File: quantum/rgb_matrix.c
Comment (XScorpion2) [reply quantum/rgb_matrix.c]: Fixed

File: quantum/rgb_matrix.c
Comment (XScorpion2) [reply quantum/rgb_matrix.c]: Fixed

File: quantum/split_common/transport.c
Comment (XScorpion2) [reply quantum/split_common/transport.c]: Fixed

File: quantum/split_common/transport.c
Comment (XScorpion2) [reply quantum/split_common/transport.c]: Fixed

File: quantum/split_common/transport.c
Comment (XScorpion2) [reply quantum/split_common/transport.c]: Fixed

File: quantum/split_common/transport.c
Comment (XScorpion2) [reply quantum/split_common/transport.c]: Fixed

File: quantum/split_common/transport.c
Comment (XScorpion2) [reply quantum/split_common/transport.c]: Fixed

File: quantum/split_common/transport.c
Comment (XScorpion2) [reply quantum/split_common/transport.c]: Fixed

File: quantum/split_common/transport.c
Comment (XScorpion2) [reply quantum/split_common/transport.c]: Fixed

**Comment by tzarc** [comment]
Needs some conflict resolution after #11046 went in.

File: quantum/split_common/transport.c
Comment (tzarc) [inline quantum/split_common/transport.c:241]: This got duped after merges and conflict resolution or something?

File: tmk_core/common/eeconfig.h
Comment (tzarc) [reply tmk_core/common/eeconfig.h:52]: Was fixed in #11046 -- I think we're going to have to decrement it again, for safety's sake.

**Review by tzarc — CHANGES_REQUESTED** [review]

**Comment by tzarc** [comment]
Modded the files to fixup the issues post-merge -- will test.

**Comment by XScorpion2** [comment]
@tzarc I just rebased, fixed up similar issues, and force pushed. You might want to re-review or re-pull and test the latest.

File: quantum/rgb_matrix.c
Comment (tzarc) [inline quantum/rgb_matrix.c:209]: This needs inversion:
```suggestion
#if !defined(RGB_MATRIX_ENABLE) || !defined(RGB_MATRIX_SPLIT)
```

**Review by tzarc — CHANGES_REQUESTED** [review]

File: quantum/rgb_matrix.c
Comment (XScorpion2) [reply quantum/rgb_matrix.c:209]: Oh I see what happened, this was when we added RGB_MATRIX_ENABLE protections to a few things. This one should not have been modified as process_rgb_matrix is only called if RGB_MATRIX_ENABLE is defined, so the check here was kinda pointless.

Comment (XScorpion2) [reply quantum/rgb_matrix.c]: reverted the change to the ifndef as that was unnecessary in the first place.

**Review by tzarc — APPROVED** [review]

**Review by drashna — APPROVED** [review]

**Comment by drashna** [comment]
Actually, one last thing.  Could the suspend status be synced between the boards?   This way, if the master side is suspended... both sides turn off.

**Comment by XScorpion2** [comment]
> 
> 
> Actually, one last thing. Could the suspend status be synced between the boards? This way, if the master side is suspended... both sides turn off.

Let me look into that this weekend.

**Comment by drashna** [comment]
Not a problem.  I am doing this locally, with a simple bool in the transport, checking the suspend status, and then setting it on the slave side.  Works well. n

**Comment by digex** [comment]
Not working transport to right side with 
`rgb_matrix_set_color(40, RGB_ORANGE)`
Work only 1 time an boot.
Animations work correctly left/right side
I use 
```
void rgb_matrix_indicators_user(void) {   
   if (host_keyboard_led_state().caps_lock) rgb_matrix_set_color(40, RGB_ORANGE );   
   else rgb_matrix_set_color(40, RGB_GREEN); 
}
```

**Comment by XScorpion2** [comment]
> 
> 
> Not working transport to right side with
> `rgb_matrix_set_color(40, RGB_ORANGE)`
> Work only 1 time an boot.
> Animations work correctly left/right side
> I use
> 
> ```
> void rgb_matrix_indicators_user(void) {   
>    if (host_keyboard_led_state().caps_lock) rgb_matrix_set_color(40, RGB_ORANGE );   
>    else rgb_matrix_set_color(40, RGB_GREEN); 
> }
> ```

Slave would need to know keyboard led state and be the one to send the color command. That is outside of the scope of this PR.

**Comment by digex** [comment]
Ok.

> Slave would need to know keyboard led state

if
```
if (IS_LAYER_ON(4)){
   for (int i = 0; i < 75; ++i) rgb_matrix_set_color(i, RGB_OFF );	   	
   	rgb_matrix_set_color(9, RGB_RED );
	rgb_matrix_set_color(15, RGB_RED );
	rgb_matrix_set_color(16, RGB_RED );
	rgb_matrix_set_color(60, RGB_RED );
	rgb_matrix_set_color(5, RGB_RED );	
    }
```

also not working, but slave and master know use LAYER (4)

**rgb_matrix_set_color()** work only master side

**Comment by XScorpion2** [comment]
> 
> 
> Ok.
> 
> > Slave would need to know keyboard led state
> 
> if
> 
> ```
> if (IS_LAYER_ON(4)){
>    for (int i = 0; i < 75; ++i) rgb_matrix_set_color(i, RGB_OFF );	   	
>    	rgb_matrix_set_color(9, RGB_RED );
> 	rgb_matrix_set_color(15, RGB_RED );
> 	rgb_matrix_set_color(16, RGB_RED );
> 	rgb_matrix_set_color(60, RGB_RED );
> 	rgb_matrix_set_color(5, RGB_RED );	
>     }
> ```
> 
> also not working, but slave and master know use LAYER (4)
> 
> **rgb_matrix_set_color()** work only master side

https://github.com/qmk/qmk_firmware/pull/11055/files#diff-f33c5d8f10ba7a970b44c1522ee1b5c5cc62243915cdfff9815068c689134298R190
The rgb_matrix_set_color handles the index conversion on slave side and master side correctly. Your code logic relies on data that does not exist on the slave side. IS_LAYER_ON for example and host_keyboard_led_state are both main side functions only and will not work as expected when run on the slave side. Which is beyond the scope of this pr.

**Comment by drashna** [comment]
@digex Just a heads up, #11930 seeks to allow you to sync additional information, in such a way that is user (and keyboard) configurable.  If you check my fork, and the "custom_drashna" branch, I have both implemented.

Additionally, I want to add more data to the sync info, such as layer state, host led status, and more. See #11749 for reference. 

but as was said, what you want is outside the scope of the PR.  Though, you can also do this by using a custom transport (check my userspace). 

@XScorpion2 did you get a chance to add the suspend state?  I think that's the only thing holding up the PR, at this point.

**Comment by XScorpion2** [comment]
@drashna didn't get to it this weekend sadly as I had to head out of town. Will try and slap it in tonight as it's simple enough. Might have to try #11930 for something specific I am working on, thanks for that!

**Comment by drashna** [comment]
No worries, life happens!

And yeah #11930 works very well, and IIRC, tzarc has added a start to multi-transactional support for i2c, as well.

**Comment by drashna** [comment]
Merge conflict.
Also, because the rgb matrix task has been moved back to quantum.c, the suspend changes aren't necessary. 

Additionally, I'm having some weird issues with this PR.  Such as the mousekeys not staying pressed, as the pointer for my trackball (handled over split comms) goes crazy, like the split data isn't synced right.

**Comment by XScorpion2** [comment]
Suspend sync and merge conflicts fixed. Tested locally on the Zygomorph split pr https://github.com/qmk/qmk_firmware/pull/11083

**Comment by tzarc** [comment]
Sorry about that, GitHub decided to delete the `develop` branch from upstream when we merged it, despite being told otherwise. Reopened.

**Comment by fauxpark** [comment]
I think I should note here that I'm currently bringing the LED Matrix feature up to par with RGB Matrix, so if that work completes before this is merged, it may be worth extending the scope of this PR to LED Matrix too :)

**Comment by tzarc** [comment]
> I think I should note here that I'm currently bringing the LED Matrix feature up to par with RGB Matrix, so if that work completes before this is merged, it may be worth extending the scope of this PR to LED Matrix too :)

If it's working I'm inclined to get it in now... it's been very close to merge for quite some time.
Will give it a test tonight.

**Review by tzarc — APPROVED** [review]

**Comment by sigprof** [comment]
Apparently this PR includes some changes that were not documented properly:

1) Calling `rgb_matrix_set_flags()` with a different value of `flags` than the previously used one will now cause all LEDs not covered by effects or `rgb_matrix_indicators_*()` to be turned off (`rgb_matrix_set_color_all(0, 0, 0);`). This is not a bad thing by itself (leaving LEDs previously used by effects in their last state was probably not useful), but still is a behavior change that might deserve to be documented (although noting it in the list of breaking changes might be enough).

2) The state of flags set by `rgb_matrix_set_flags()` is now stored in the `rgb_matrix_config` structure, which is saved in the EEPROM and gets restored after a restart. However, unlike all other `rgb_matrix_set_*()` functions without `_noeeprom` in the name, `rgb_matrix_set_flags()` only set the `.flags` field in RAM, but does not trigger a write to the EEPROM when the value has been changed. But if at some later time another `rgb_matrix_set_*()` function is called and triggers a write to the EEPROM, the value of `.flags` will also be saved along with the rest of `rgb_matrix_config`, and will get restored after a subsequent restart. So `rgb_matrix_set_flags()` behaves like some `_noeeprom` function would, but does not have `_noeeprom` in its name.

**Comment by fpazos** [comment]
Hi, I've been trying this on a lily58l, despite light effects going through outer ends of the board (maybe due to my changes), it seems to be working fine. Nice feature, thanks!

**Pull request merged**
